File: npmjs/pray-cli/src/resolve/git-refresh.ts
import { existsSync } from "node:fs";
import { PrayError } from "../errors.js";
import { readLockfile } from "../lockfile/index.js";
import type { Lockfile } from "../lockfile/types.js";

export function resolutionMayBenefitFromGitSourceRefresh(
  error: unknown,
): boolean {
  return (
    error instanceof PrayError &&
    error.kind === "resolution" &&
    catalogMissMessage(error.message)
  );
}

function catalogMissMessage(message: string): boolean {
  return (
    message.includes("no registry version") ||
    message.includes("not found in distribution") ||
    message.includes("not found in git source") ||
    message.includes("v1/packages/")
  );
}

export function annotateMissingGitCatalog(
  error: unknown,
  packageName: string,
  sourceName: string,
  revision: string,
): unknown {
  if (
    !(error instanceof PrayError) ||
    error.kind !== "resolution" ||
    revision.length === 0 ||
    !catalogMissMessage(error.message) ||
    error.message.includes("not found in git source")
  ) {
    return error;
  }
  return PrayError.resolution(
    `package ${packageName} was not found in git source ${sourceName} at revision ${revision}. ` +
      "Run `pray update` to advance the source pin.",
  );
}

export function annotateFailedGitRefresh(
  lockfilePath: string,
  error: unknown,
): unknown {
  if (
    !(error instanceof PrayError) ||
    error.kind !== "resolution" ||
    !catalogMissMessage(error.message)
  ) {
    return error;
  }
  const pins = lockedGitRevisionGuidance(lockfilePath);
  if (pins === undefined) {
    return error;
  }
  const packageName =
    packageNameFromCatalogMiss(error.message) ?? "the declared package";
  return PrayError.resolution(
    `package ${packageName} was not found in locked git source (${pins}). ` +
      "Run `pray update` to advance the source pin.",
  );
}

function lockedGitRevisionGuidance(lockfilePath: string): string | undefined {
  if (!existsSync(lockfilePath)) {
    return undefined;
  }
  let lockfile: Lockfile;
  try {
    lockfile = readLockfile(lockfilePath);
  } catch {
    return undefined;
  }
  const pins = lockfile.source
    .filter((source) => source.kind === "git" && source.revision)
    .map((source) => `${source.name} at ${source.revision}`);
  return pins.length === 0 ? undefined : pins.join(", ");
}

function packageNameFromCatalogMiss(message: string): string | undefined {
  if (!message.startsWith("package ")) {
    return undefined;
  }
  const rest = message.slice("package ".length);
  const ending =
    rest.indexOf(" was not found") === -1
      ? rest.indexOf(" not found")
      : rest.indexOf(" was not found");
  if (ending === -1) {
    return undefined;
  }
  const name = rest.slice(0, ending).trim();
  return name.length === 0 ? undefined : name;
}

File: rubygems/pray-cli/lib/pray/git_refresh.rb
# frozen_string_literal: true

module Pray
  module GitRefresh
    module_function

    def resolution_may_benefit_from_git_source_refresh?(error)
      error.category == :resolution && catalog_miss_message?(error.message)
    end

    def catalog_miss_message?(message)
      message.include?("no registry version") ||
        message.include?("not found in distribution") ||
        message.include?("not found in git source") ||
        message.include?("v1/packages/")
    end

    def annotate_missing_git_catalog(error, package_name:, source_name:, revision:)
      return error unless error.category == :resolution
      return error if revision.to_s.empty?
      return error unless catalog_miss_message?(error.message)
      return error if error.message.include?("not found in git source")

      Error.resolution(
        "package #{package_name} was not found in git source #{source_name} at revision #{revision}. " \
        "Run `pray update` to advance the source pin."
      )
    end

    def annotate_failed_refresh(lockfile_path, error)
      return error unless error.category == :resolution
      return error unless catalog_miss_message?(error.message)

      pins = locked_git_revision_guidance(lockfile_path)
      return error if pins.nil?

      package = package_name_from_catalog_miss(error.message) || "the declared package"
      Error.resolution(
        "package #{package} was not found in locked git source (#{pins}). " \
        "Run `pray update` to advance the source pin."
      )
    end

    def locked_git_revision_guidance(lockfile_path)
      return unless File.exist?(lockfile_path)

      lockfile = LockfileIO.read_lockfile(lockfile_path)
      pins = lockfile.source.filter_map do |source|
        next unless source.kind == "git" && source.revision

        "#{source.name} at #{source.revision}"
      end
      pins.empty? ? nil : pins.join(", ")
    rescue Error
      nil
    end

    def package_name_from_catalog_miss(message)
      rest = message.delete_prefix("package ")
      return if rest == message

      ending = rest.index(" was not found") || rest.index(" not found")
      return unless ending

      name = rest[0, ending].strip
      name.empty? ? nil : name
    end
  end
end
